engines/gargoyle/detection.cpp

```diff
@@ -0,0 +1,323 @@
+		*engine = new Gargoyle::Frotz::Frotz(syst, gd);
+		break;
+	case Gargoyle::INTERPRETER_SCOTT:
+		*engine = new Gargoyle::Scott::Scott(syst, gd);
+		break;
+	default:
+		error("Unknown interpreter");
+	}
+
+	return gd != 0;
+}
+
+SaveStateList GargoyleMetaEngine::listSaves(const char *target) const {
+	Common::SaveFileManager *saveFileMan = g_system->getSavefileManager();
+	Common::StringArray filenames;
+	Common::String saveDesc;
+	Common::String pattern = Common::String::format("%s.0##", target);
+	Gargoyle::SavegameHeader header;
+
+	filenames = saveFileMan->listSavefiles(pattern);
+
+	SaveStateList saveList;
+	for (Common::StringArray::const_iterator file = filenames.begin(); file != filenames.end(); ++file) {
+		const char *ext = strrchr(file->c_str(), '.');
+		int slot = ext ? atoi(ext + 1) : -1;
+
+		if (slot >= 0 && slot <= MAX_SAVES) {
+			Common::InSaveFile *in = g_system->getSavefileManager()->openForLoading(*file);
+
+			if (in) {
+				if (Gargoyle::FileStream::readSavegameHeader(in, header))
+					saveList.push_back(SaveStateDescriptor(slot, header._saveName));
+
+				delete in;
+			}
+		}
+	}
+
+	// Sort saves based on slot number.
+	Common::sort(saveList.begin(), saveList.end(), SaveStateDescriptorSlotComparator());
+	return saveList;
+}
+
+int GargoyleMetaEngine::getMaximumSaveSlot() const {
+	return MAX_SAVES;
+}
+
+void GargoyleMetaEngine::removeSaveState(const char *target, int slot) const {
+	Common::String filename = Common::String::format("%s.%03d", target, slot);
+	g_system->getSavefileManager()->removeSavefile(filename);
+}
+
+SaveStateDescriptor GargoyleMetaEngine::querySaveMetaInfos(const char *target, int slot) const {
+	Common::String filename = Common::String::format("%s.%03d", target, slot);
+	Common::InSaveFile *in = g_system->getSavefileManager()->openForLoading(filename);
+
+	if (in) {
+		Gargoyle::SavegameHeader header;
+		if (Gargoyle::FileStream::readSavegameHeader(in, header)) {
+			// Create the return descriptor
+			SaveStateDescriptor desc(slot, header._saveName);
+			desc.setSaveDate(header._year, header._month, header._day);
+			desc.setSaveTime(header._hour, header._minute);
+			desc.setPlayTime(header._totalFrames * GAME_FRAME_TIME);
+
+			delete in;
+			return desc;
+		}
+	}
+
+	return SaveStateDescriptor();
+}
+
+DetectedGames GargoyleMetaEngine::detectGames(const Common::FSList &fslist) const {
+	DetectedGames detectedGames;
+	Gargoyle::Frotz::FrotzMetaEngine::detectGames(fslist, detectedGames);
+	Gargoyle::Scott::ScottMetaEngine::detectGames(fslist, detectedGames);
+
+	return detectedGames;
+}
+
+static Gargoyle::GargoyleGameDescription gameDescription;
+
+ADDetectedGames GargoyleMetaEngine::detectGame(const Common::FSNode &parent, const FileMap &allFiles, Common::Language language, Common::Platform platform, const Common::String &extra) const {
+	static char gameId[100];
+	strcpy(gameId, ConfMan.get("gameid").c_str());
+	Common::String filename = ConfMan.get("filename");
```

Comment:
I'm not so sure about storing the filename in the configuration file when detecting the game and then using it when launching. That does not work with the command line interface where you can start a game just from a gameid and a path, without a target / bypassing the detection code (`scummvm --path=~/games/IF/Scott/ adventureland`).

Reply:
That may indeed be a problem. But I'm not sure if it's avoidable. Currently, adventures only tend to have a single adventure per folder, with a known set of files. Or, in rare cases like DOTT, technically two games, but still with a distinct set of known files. For the case of Interactive Fiction, it's more problematic, because dozens of games may end up being in the same folder, including multiple versions of the same game.

For example, I have a folder structure with a folder for each game, and within each all the different versions of that game (or at least, those that I've collected). If you look at https://github.com/dreammaster/scummvm/blob/gargoyle/engines/gargoyle/frotz/detection_tables.cpp  you'll see that I have 5 different versions of The Hitchhiker's Guide to the Galaxy. If I used the command line like in your example, scummvm --path=~/games/if/hhgttg/ hhgttg   how would it know which version of the game to launch. It gets even worse with popular interactive fiction formats like zcode, where there are hundreds, maybe thousands, of fan games which could all be present in the one folder.

I'm not really familiar with the detection code too much. In all the cases in the past, it was easy simple to use AdvancedMetaEngine with a hardcoded set of files/sizes/md5s, so I didn't really look into how it did it's magic. :). I don't suppose the path can optionally specify a specific filename as well? Spitballing ideas.. even if it doesn't already, maybe it would be something better to do as a change to the advanced meta engine.. in that when a path specifies a specific filename, it would transparently treat it as if in the parent directory with only that one file passed to the detectGames method. That way, I wouldn't need a separate filename field. It still doesn't really do away with the messy startup code I currently need for the startup, though.